xdp/src/packet.rs

```diff
@@ -0,0 +1,116 @@
+#![allow(clippy::arithmetic_side_effects)]
+
+use {libc::ETH_P_IP, std::net::Ipv4Addr};
+
+pub const ETH_HEADER_SIZE: usize = 14;
+pub const IP_HEADER_SIZE: usize = 20;
+pub const UDP_HEADER_SIZE: usize = 8;
+
+pub fn write_eth_header(packet: &mut [u8], src_mac: &[u8; 6], dst_mac: &[u8; 6]) {
+    packet[0..6].copy_from_slice(dst_mac);
+    packet[6..12].copy_from_slice(src_mac);
+    packet[12..14].copy_from_slice(&(ETH_P_IP as u16).to_be_bytes());
+}
+
+pub fn write_ip_header(packet: &mut [u8], src_ip: &Ipv4Addr, dst_ip: &Ipv4Addr, udp_len: u16) {
+    let offset = ETH_HEADER_SIZE;
+    let total_len = IP_HEADER_SIZE + udp_len as usize;
+
+    // version (4) and IHL (5)
+    packet[offset] = 0x45;
+    // tos
+    packet[offset + 1] = 0;
+    packet[offset + 2..offset + 4].copy_from_slice(&(total_len as u16).to_be_bytes());
+    // identification
+    packet[offset + 4..offset + 6].copy_from_slice(&0u16.to_be_bytes());
+    // flags & frag offset
+    packet[offset + 6..offset + 8].copy_from_slice(&0u16.to_be_bytes());
+    // TTL
+    packet[offset + 8] = 64;
+    // protocol (UDP = 17)
+    packet[offset + 9] = 17;
+    // checksum
+    packet[offset + 10..offset + 12].copy_from_slice(&0u16.to_be_bytes());
+    packet[offset + 12..offset + 16].copy_from_slice(&src_ip.octets());
+    packet[offset + 16..offset + 20].copy_from_slice(&dst_ip.octets());
+
+    let checksum = calculate_ip_checksum(&packet[offset..offset + IP_HEADER_SIZE]);
+    packet[offset + 10..offset + 12].copy_from_slice(&checksum.to_be_bytes());
+}
+
+pub fn write_udp_header(
+    packet: &mut [u8],
+    src_ip: &Ipv4Addr,
+    src_port: u16,
+    dst_ip: &Ipv4Addr,
+    dst_port: u16,
+    payload_len: u16,
+    csum: bool,
+) {
+    let offset = ETH_HEADER_SIZE + IP_HEADER_SIZE;
+    let udp_len = UDP_HEADER_SIZE + payload_len as usize;
+
+    packet[offset..offset + 2].copy_from_slice(&src_port.to_be_bytes());
+    packet[offset + 2..offset + 4].copy_from_slice(&dst_port.to_be_bytes());
+    packet[offset + 4..offset + 6].copy_from_slice(&(udp_len as u16).to_be_bytes());
+    packet[offset + 6..offset + 8].copy_from_slice(&0u16.to_be_bytes());
+
+    if csum {
```

Comment:
Then checksum would be a separate operation (as we may want to offload this to the nic)

Reply:
It is already a separate operation, and if you see where this is called, we indeed don't compute it.